### PR TITLE
Samples: remove internal headers

### PR DESCRIPTION
**Pull request by antkmsft**
Samples:
* remove `#include http_internal.h`
* init `az_keyvault_create_key_options` properly (`az_keyvault_create_key_options_default()`)
* remove unused headers (json in storage sample)
* remove unused/almost unused exit_code - that's an unnecessary complication if we don't use it.

We need to ask someone more knowledgeable in CMake to help us to fix our CMake files, so that even if you `#include <az_http_internal.h>` in sample.c, it won't find that header and won't compile.

Also, what do you think of `TEST_FAIL_ON_ERROR` macro? I am not touching it in this PR.
* Why is the name `TEST`? I think `EXIT_ON_ERROR` would be a better name.
* Is it a neccessary concept for users to bring their attention to and to understand?
In addition to some natural extra concept understanding, it has `do { ... } while(0)`, which is also a useful technique to use in our library code, but in a sample? Well, if I am a beginner, I see do-while-0, and I have questions already.
Do we want our users to replicate exactly that approach in their code? Why not use `if (failed()) { printf(); return 1; }`?

I have left all this feedback at a time in the PR, and in Teams. If you disagree with my proposal, it is fine, but if you strongly agree, then please let me know how I can bring more attention to these comments next time, block the PR?

If having `TEST_FAIL_ON_ERROR` is rather necessary for us in order to test that samples do not fail in CI/CD, we still can have sample code written as
```
if (az_failed(...))
{
    printf(...);
    exit 1;
}
```
and we should be able to test it in CI/CD.
First, we can analyze the exit code, it is only non-0 when something failed (and also if code was compiled with az_nohttp, and I think that we should return non-0 in that case too).
Second, another option would be to define a macro in only in our CI/CD, and name it exactly as the function, `az_failed(x)` and in that macro, if the result is faulty, do something extra that we need in CI/CD.

Update: here is the proposal to remove `TEST_FAIL_ON_ERROR_MACRO`: https://github.com/Azure/azure-sdk-for-c/pull/758

### CODE REVIEW

**Comment by gilbertw** [comment]
The do while in the MACRO is a very common thing to do C/C++.  We could add a note in a readme linking to the many stackoverflow threads on the subject.

File: sdk/samples/keyvault/keyvault/samples/src/keys_client_example.c [hunk 1]
```diff
@@ -43,7 +43,6 @@
 #include <az_context.h>
 #include <az_credentials.h>
 #include <az_http.h>
-#include <az_http_internal.h>
 #include <az_json.h>
 #include <az_keyvault.h>
```
Comment (gilbertw) [inline sdk/samples/keyvault/keyvault/samples/src/keys_client_example.c:48]: RETURN_IF_FAILED(.......) is a better name for the TEST_FAIL_ON_ERROR macro

Comment (antkmsft) [reply sdk/samples/keyvault/keyvault/samples/src/keys_client_example.c:48]: Here is what I propose: https://github.com/Azure/azure-sdk-for-c/pull/758

**Comment by antkmsft** [comment]
ALSO SEE https://github.com/Azure/azure-sdk-for-c/pull/758

**Comment by ahsonkhan** [comment]
> We need to ask someone more knowledgeable in CMake to help us to fix our CMake files, so that even if you `#include <az_http_internal.h>` in sample.c, it won't find that header and won't compile.

Hopefully @barcharcraz can provide some guidance here. We want to make sure samples aren't using internal headers.

**Comment by ahsonkhan** [comment]
> I have left all this feedback at a time in the PR, and in Teams. If you disagree with my proposal, it is fine, but if you strongly agree, then please let me know how I can bring more attention to these comments next time, block the PR?

I definitely agree with the suggestions, at least the spirit of it. We can hash out the specific details in #758.

I think leaving blocking feedback or talking through the feedback points in the GitHub PR is a perfectly reasonable thing to do, and I would think about it on a case-by-case basis. If it is truly blocking, i.e. bug, or bad code structure/design, or code smells, I convey that in my code reviews and expect/hope others do the same. If it is contentious (with differing opinions) or a time sensitive change, filing issues and deferring is OK too.

**Comment by vhvb1989** [comment]
> > We need to ask someone more knowledgeable in CMake to help us to fix our CMake files, so that even if you `#include <az_http_internal.h>` in sample.c, it won't find that header and won't compile.
> 
> Hopefully @barcharcraz can provide some guidance here. We want to make sure samples aren't using internal headers.

I don't think that is possible, since we need to expose those for services (keyvault, storage, iot)
The way we did this in the past was with az_core_internal lib containing only internal stuff and a the lib that services will include but not customers

**Comment by antkmsft** [comment]
@ahsonkhan , @gilbertw , @vhvb1989 , you approved this PR, but I have a relatively big change - we no longer `#include <_az_cfg.h>` in sample code (just single commit: https://github.com/Azure/azure-sdk-for-c/pull/757/commits/29c5dea22217c9a8c8e9e487fb6a5de7cc2120c5 , but I recommnd to review the whole PR again, as I added your feedback and some of my fixes from the other PR #758 (didn't remove the macro though, as you didn't agree on that)).

**Comment by antkmsft** [comment]
I haven't heard new feedback, but I am open to it. If there's something you'd like to see changed or clarified, feel free to contact me or leave comments in this PR, we may create a new one for more changes, but this PR I am no longer holding off and checking in.

File: sdk/core/core/inc/az_span.h [hunk 2]
```diff
@@ -54,10 +54,39 @@ AZ_NODISCARD AZ_INLINE int32_t az_span_size(az_span span) { return span._interna
...
+
+// Same function as above, only for use in AZ_SPAN_FROM_BUFFER macro.
+// So that the users don't get warnings in their code when they use the macro:
+// C4221: nonstandard extension: 'ptr': cannot be initialized using address of automatic variable
+// C4204: nonstandard extension: non-constant aggregate initializer
+AZ_NODISCARD AZ_INLINE az_span _az_span_init(uint8_t* ptr, int32_t size)
+{
+  return (az_span){ ._internal = { .ptr = ptr, .size = size } };
+}
+
```
Comment (ahsonkhan) [inline sdk/core/core/inc/az_span.h:83]: I wish we didn't have to do this. Can we workaround the warning in a different way that doesn't require code duplication? Who is seeing these warnings because when I build locally, I didn't. Is  this a side effect of removing `<_az_cfg.h>`?

In the future, please split changes like these out into separate PRs.

File: sdk/core/core/inc/az_span.h [hunk 3]
```diff
@@ -401,7 +407,12 @@ typedef struct
  *
  */
 #define AZ_PAIR_NULL \
-  (az_pair) { .key = AZ_SPAN_NULL, .value = AZ_SPAN_NULL }
+  (az_pair) \
+  { \
+    .key = { ._internal = { .ptr = NULL, .size = 0 } }, \
+    .value \
+        = {._internal = { .ptr = NULL, .size = 0 } } \
```
Comment (ahsonkhan) [inline sdk/core/core/inc/az_span.h:414]: Why was this change necesssary?

File: sdk/core/core/src/az_span.c [hunk 4]
```diff
@@ -584,7 +584,7 @@ AZ_NODISCARD static az_span _az_span_trim_side(az_span source, az_span_trim_side
   if (side == RIGHT)
   {
     increment = -1; // Set increment to be decremental for moving ptr
-    source_ptr += (source_size - 1); // Set initial position to the end
+    source_ptr += ((size_t)source_size - 1); // Set initial position to the end
```
Comment (ahsonkhan) [inline sdk/core/core/src/az_span.c:587]: We don't cast to `size_t` anywhere else. Why was this necessary?

File: sdk/samples/keyvault/keyvault/samples/src/keys_client_example.c [hunk 5]
```diff
@@ -69,57 +66,67 @@
...
 az_span get_key_version(az_http_response* response);
 az_span const key_name_for_test = AZ_SPAN_LITERAL_FROM_STR("test-new-key");
 
+#ifdef _MSC_VER
+// "'getenv': This function or variable may be unsafe. Consider using _dupenv_s instead."
+#pragma warning(disable : 4996)
+// "Compiler will insert Spectre mitigation for memory load if /Qspectre switch specified"
+#pragma warning(disable : 5045)
+#endif // _MSC_VER
+
```
Comment (ahsonkhan) [inline sdk/samples/keyvault/keyvault/samples/src/keys_client_example.c:78]: It isn't great that we have to disable such warnings inside the samples. I don't have a great alternative though, atm.

Comment (antkmsft) [reply sdk/samples/keyvault/keyvault/samples/src/keys_client_example.c:78]: Other options were more verbose and were drawing too much attention, which wasn't good for samples. I don't like this either, and I tried to avoid this.

File: sdk/core/core/src/az_span.c (hunk 4 above)
Comment (antkmsft) [reply sdk/core/core/src/az_span.c:587]: There was warning on my dogfood MSVC.

File: sdk/core/core/inc/az_span.h (hunk 3 above)
Comment (antkmsft) [reply sdk/core/core/inc/az_span.h:414]: There was a warning, and this is the way to avoid it. I'll get the specific warning and post it here.

File: sdk/core/core/inc/az_span.h (hunk 2 above)
Comment (antkmsft) [reply sdk/core/core/inc/az_span.h:83]: This is absolutely a side effect of removing az_cfg.h, where we disable a bunch of nonsensical warnings, but we can't require our customers to include az_cfg, or to disable them.
The warnings that are explicitly disabled in samples, do not come from the SDK code, but they are from the sample code.

Comment (antkmsft) [reply sdk/core/core/inc/az_span.h:83]: Regarding code duplication, I talked with Jeff, I will change the macro to simply use az_span_init(), which will check preconditions when built with precondition support, but he is fine with it, and thinks it even might help to spot some bug.
